clang/lib/AST/PreorderAST.cpp

```diff
@@ -56,11 +83,65 @@ void PreorderAST::Create(Expr *E, Node *Parent) {
   E = Lex.IgnoreValuePreservingOperations(Ctx, E->IgnoreParens());
 
   if (const auto *BO = dyn_cast<BinaryOperator>(E)) {
-    auto *N = new BinaryNode(BO->getOpcode(), Parent);
+    BinaryOperator::Opcode BinOp = BO->getOpcode();
+    Expr *LHS = BO->getLHS();
+    Expr *RHS = BO->getRHS();
+
+    // We can convert (e1 - e2) to (e1 + -e2) if -e2 does not overflow.  One
+    // instance where -e2 can overflow is if e2 is INT_MIN. Here, instead of
+    // specifically checking whether e2 is INT_MIN, we add a unary minus to e2
+    // and then check if the resultant expression -e2 overflows. If it
+    // overflows, we undo the unary minus operator.
+
+    // TODO: Currently we only handle detection of overflow for integer
+    // constant expressions. We need to handle the case where e2 is a variable
```

Comment:
I think the last sentence in this comment could be clarified.  It is kind of confusing.   We're unlikely to be able to prove that a non-constant expression overflows.  However, we could prove it does not overflow.   So I'd like us to state that we can only prove the integer constant expressions don't overflow.  We still need to handle proving that non-constant expressions do no overflow.